### ISSUE TITLE
Amy's Echo NG: greenfield rebuild for local-first personalized gesture-to-speech AAC

### ISSUE THREAD

**Issue by voku** [voku/AmysEcho#1176]
## Goal

Track the greenfield design and implementation of **Amy's Echo NG** as a local-first AAC system that uses personalized gesture recognition as one input path for communication.

This is **not** a migration project.

- no backwards compatibility
- no legacy data migration
- no NPZ compatibility
- no Base64 model transport
- no `window.*` ML hooks
- no requirement to preserve the current server/training architecture
- current AmysEcho code is reference material only: reuse proven ideas, not legacy constraints

## Product target

Amy's Echo NG should let a child communicate through personalized and DGS-adjacent gestures, with output as text, symbol/photo and speech.

The primary success metric is not classifier accuracy in isolation. It is:

> A child can successfully communicate an intended message with minimal repetition, correction or caregiver intervention.

Core product capabilities:

- continuous camera-based gesture recognition
- explicit `unknown` / no-gesture handling instead of guessing
- personalized gestures that become locally usable immediately
- local-first / offline live recognition
- text + symbol/photo + speech output
- symbol-board / visual-scene fallback
- caregiver-guided gesture teaching / recording flow
- profile isolation, export and deletion

## Target architecture

```text
Camera
  ↓
Web Worker
  ↓
MediaPipe landmarks
  ├─ hands
  ├─ reduced pose anchors
  └─ mouth-heavy face blendshapes + head pose
  ↓
Feature extraction
  ├─ normalized xyz landmarks
  ├─ geometry-aware joint angles / distances
  ├─ motion / timing features
  └─ presence / quality masks
  ↓
Multi-task temporal encoder
  ├─ gesture phase / spotting
  ├─ embedding
  └─ global classification
  ↓
Profile prototype matcher
  ↓
Decision policy
  ↓
Communication concept
  ├─ speech
  ├─ symbol/photo
  └─ context
```

### Runtime decisions

- React + TypeScript + Vite PWA
- MediaPipe vision tasks in a dedicated worker
- PyTorch for training
- ONNX export
- ONNX Runtime Web with **Wasm as the production execution path**
- IndexedDB for structured local state
- OPFS / binary storage for models and larger local artifacts
- no raw-video persistence by default

### Backend decisions

- PHP 8.x + Symfony
- MariaDB
- server-side cookie sessions
- Python ML worker for training/export
- backend is **not** in the critical live-recognition path

### Perception

Use MediaPipe landmarks rather than RGB video as the first production representation.

Feature contract includes:

- both hands with normalized xyz landmarks
- hand-presence and handedness confidence
- geometry-aware hand features such as finger/joint angles and hand distances
- reduced upper-body anchors
- mouth-focused face blendshapes and head rotation
- temporal deltas, motion and quality signals

### Temporal model

Use one compact multi-task temporal convolutional encoder.

Outputs:

- framewise gesture phase: `idle`, `start`, `active`, `end`, `invalid`
- sequence embedding
- global label logits

### Personalization

Do **not** train a full neural network per profile.

Use:

- one shared encoder
- profile-specific gesture prototypes in embedding space
- episodic / prototypical loss during training so the embedding is optimized for few-shot personalization
- multiple prototypes per personal gesture when recordings form distinct natural clusters

### Gesture spotting

Continuous recognition is a first-class problem, not a post-processing afterthought.

The runtime must explicitly model:

- idle
- start
- active
- end
- incomplete / invalid
- cooldown / duplicate suppression

### SignBridge
https://github.com/AdhithyaVar/SignBridge-Real-Time-ISL-Translator-with-Edge-Deployment

High-value ideas:

- recording state machine
- wrist-relative / scale normalization
- landmark augmentation
- multi-signer dataset organization
- LOSO-style honest evaluation
- PyTorch → ONNX numerical validation
- CPU benchmarking
- inference post-processing: smoothing → gating → cooldown → output

Do not copy its large CNN-LSTM-attention model as the target architecture.

### Few-shot Prototypical Network
https://github.com/Hadeeresameldin/ArSL-FewShot-Prototypical-Network

High-value ideas:

- episodic training
- N-way K-shot evaluation
- profile/signer-independent embedding quality
- prototypical classification under limited data

### sign-language-translator
https://github.com/sign-language-translator/sign-language-translator

High-value ideas:

- landmark-sequence domain model
- pose / landmark transformations
- deterministic augmentation pipeline
- separation of gesture/sign representation from language/text meaning

### OpenHands
https://github.com/AI4Bharat/OpenHands

High-value idea:

- pose-based pretraining across sign-language datasets as a future source for the shared encoder

### Cboard
https://github.com/cboard-org/cboard

High-value AAC ideas:

- browser/PWA communication board
- symbol categories and editable boards
- speech synthesis abstraction
- offline-first operation
- symbol ecosystems such as ARASAAC / Mulberry / Global Symbols

License constraints must be reviewed before reusing code or assets.

## Evaluation rules

Do not trust random sample splits from overlapping clips or the same recording session.

Required evaluation dimensions:

- session holdout
- day holdout
- device / camera-condition holdout
- signer-independent evaluation for the shared encoder
- profile-specific evaluation for personalization
- `unknown` recall
- false positives per minute
- duplicate outputs per real gesture
- time from gesture end to communication output
- corrections / repetitions per communication attempt

The product gate must include real communication sessions, not only offline ML metrics.

## Security / privacy constraints

- no default production secrets
- no internal filesystem paths in responses
- no raw landmarks in telemetry
- no raw-video persistence by default
- strict profile isolation
- artifact checksum validation before activation
- bounded upload sizes and schema validation
- complete profile export and deletion
- deterministic local cache invalidation on profile deletion / logout

### M1 — Product + data contract

Deliver:

- canonical communication concepts / labels
- recording protocol
- target-device matrix
- privacy rules
- feature contract v1
- runtime and product acceptance gates

### M2 — Local vertical slice

Deliver:

- PWA
- camera
- worker
- MediaPipe landmarks
- feature extraction
- ONNX Runtime Web / Wasm
- local model cache
- 3–5 real gestures end-to-end
- text / symbol / speech output

### M3 — Continuous gesture spotting

Deliver:

- phase prediction / state machine
- no-gesture handling
- incomplete-gesture rejection
- cooldown and duplicate suppression
- quality gating

### M4 — Training pipeline

Deliver:

- dataset model
- deterministic augmentation
- grouped/session-aware splits
- temporal encoder training
- episodic prototype loss
- ONNX export
- PyTorch/ONNX parity checks

### M5 — Personalization

Deliver:

- caregiver recording flow
- immediate local prototype creation
- conflict / outlier detection
- multiple prototypes per gesture where justified
- offline use immediately after recording

### M6 — AAC product layer

Deliver:

- communication concept model
- symbol/photo output
- editable board categories
- visual scenes
- recent/favorites
- caregiver teaching flow
- local speech synthesis

### M7 — Backend + sync

Deliver:

- Symfony API
- MariaDB
- profile authz
- example/correction sync
- training jobs
- immutable model artifacts
- export + deletion

### M8 — Product hardening

Deliver:

- long-session tests
- offline/reconnect
- low-light / background variance
- camera failures
- storage-pressure behavior
- profile switching
- artifact corruption handling
- real-world communication sessions

## Definition of Done

Amy's Echo NG 1.0 is done when:

- the agreed core gestures work on the target devices
- live recognition works offline
- uncertain gestures are rejected instead of guessed
- duplicate output from a held gesture is controlled
- personal gestures can be added from a small guided recording set without retraining a full profile model
- personal gestures are immediately usable locally
- text, symbol/photo and speech output all work
- the AAC board remains available as an immediate fallback
- profile data remains isolated
- export and deletion are complete
- runtime, ML, privacy and real-world communication gates all pass

## First concrete slice

Start with **M1 + the smallest M2 vertical slice**:

1. define 3–5 representative real gestures
2. define the first feature contract
3. implement worker-based landmark capture
4. store deterministic landmark sequences
5. run a tiny temporal encoder through PyTorch → ONNX → ONNX Runtime Web/Wasm
6. prove one full offline path from camera to communication output

Do not build backend, sync, dashboards or generalized model management before this local vertical slice works.
